### ISSUE TITLE
Support app clips on iOS 14

### ISSUE THREAD

**Issue by xster** [flutter/flutter#60029]
We'll want more details from https://developer.apple.com/videos/play/wwdc2020/10174/. So far, it seems like it's just a separate build target in Xcode and you choose which target your sources get compiled into. Hopefully that includes selection for frameworks too.

**Comment by EliasDeuss** [comment]
Would that work because the limit for a clip is 10mb (if I remember correctly) and flutters framework is more than that. I could be wrong though. @xster

**Comment by jonahwilliams** [comment]
The default demo app is about 5.4 MB, that is plenty of room for small applications

**Comment by NoahJApps** [comment]
I'm planning on just exposing my flutter data in swift and then learning swift UI for the widgets... I doubt Google can get the Flutter Engine small enough to fit 10mb

**Comment by vincentracine** [comment]
As the adoption of NFC increases, this feature will become a must-have.

**Comment by j-mateo** [comment]
Would be great if it is cross platform. Clips for iOS,  and [Instant Apps](https://developer.android.com/topic/google-play-instant/) for Android?

**Comment by marqueeeeeee** [comment]
it looks like it is possible based from the initial documents and it is similar to universal links.

**Comment by xster** [comment]
right, the devil's in the details but from https://developer.apple.com/videos/play/wwdc2020/10174/, it seems so far like there aren't any blocking conceptual issues.

**Comment by chris-makaio** [comment]
well, exporting thin builds most of them show around 10MB for the full app. Given that, I assume that a size-optimized target for App Clip including only necessary features should be well below 10 MB. I guess we'd just need a mechanism for the new target, right?

**Comment by anandat** [comment]
hmm right as per the app clip example given by apple ex : fruta

**Comment by Isakdl** [comment]
Really cool feature, would love to implement this in our app! These issues would probably help make this easier / viable https://github.com/flutter/flutter/issues/16815 https://github.com/flutter/flutter/issues/16813

**Comment by jmagman** [comment]
I was able to create an App Clip running the Flutter demo by opening `ios/Runner.xcworkspace` in Xcode 12 and manually adding the target.  This is not for the faint of heart and requires some Xcode knowledge to pull off.  

1. Open `ios/Runner.xcworkspace`
1. Open the Project settings and hit the + to add a new target
![Screen Shot 2020-06-30 at 3 07 25 PM](https://user-images.githubusercontent.com/682784/86183076-12455700-bae6-11ea-85ed-f4cb86c23839.png)
1. Add an App Clip Target
![Screen Shot 2020-06-30 at 3 08 52 PM](https://user-images.githubusercontent.com/682784/86183110-28ebae00-bae6-11ea-96cb-9b26195e21b0.png)
1. Interface: Storyboard, Life Cycle: UI App Delegate
![Screen Shot 2020-06-30 at 3 09 05 PM](https://user-images.githubusercontent.com/682784/86183178-4c165d80-bae6-11ea-9223-3bd9a59b43e6.png)
1. When prompted, hit Activate.
![Screen Shot 2020-06-30 at 3 10 12 PM](https://user-images.githubusercontent.com/682784/86183187-50db1180-bae6-11ea-955f-2260b39f5789.png)
1. Delete the newly created AppDelegate.swift, ViewController.swift, Main.storyboard, Assets.xcassets, LaunchScreen.storyboard
![Screen Shot 2020-06-30 at 3 10 43 PM](https://user-images.githubusercontent.com/682784/86183273-841da080-bae6-11ea-86c4-c6fa5087ec7c.png)
![Screen Shot 2020-06-30 at 3 10 59 PM](https://user-images.githubusercontent.com/682784/86183297-8f70cc00-bae6-11ea-9ca0-55faa21367bd.png)
1. Open Project (not Target!) settings > Info > Deployment Configurations. Base the new App Clip target on the same configurations as your Flutter Runner target.  This will give them access to all the Flutter build settings (build number from your pubspec.yaml, etc). So "None" will become either "Debug" or "Release".  Set the iOS Deployment Target to 13 (I couldn't get this working on < 13).
![Screen Shot 2020-06-30 at 3 11 31 PM](https://user-images.githubusercontent.com/682784/86184296-be883d00-bae8-11ea-8685-2de2e8abcf8e.png)
1. Open the App Clip entitlements file.  Add the [App Clip Associated Domains](https://developer.apple.com/documentation/app_clips/creating_an_app_clip#3604097).
![Screen Shot 2020-06-30 at 3 12 54 PM](https://user-images.githubusercontent.com/682784/86183494-f8f0da80-bae6-11ea-87d3-88bba596b4a0.png)
1. Compile/embed your exiting Runner Swift files and assets into both Runner and the new App Clip target by setting the membership of each of the following files:
AppDelegate.swift, ViewController.swift, Main.storyboard, Assets.xcassets, LaunchScreen.storyboard
![Screen Shot 2020-06-30 at 3 13 14 PM](https://user-images.githubusercontent.com/682784/86183772-9ba95900-bae7-11ea-92fc-36a19ca3cf43.png)
![Screen Shot 2020-06-30 at 3 13 10 PM](https://user-images.githubusercontent.com/682784/86183774-9d731c80-bae7-11ea-894c-8dab1d55482d.png)
![Screen Shot 2020-06-30 at 3 13 07 PM](https://user-images.githubusercontent.com/682784/86183775-9e0bb300-bae7-11ea-8e95-0debe3050961.png)
etc
1. Open the App Clip Info.plist.  Set `Build version string (short)` to `$(FLUTTER_BUILD_NAME)` and `Bundle version` to `$(FLUTTER_BUILD_NUMBER)`.  These need to match between the app and app clip target.
![Screen Shot 2020-06-30 at 3 13 43 PM](https://user-images.githubusercontent.com/682784/86183871-cdbabb00-bae7-11ea-8538-0709fcf2278a.png)
1. Create a new entitlements file for your Runner app.  Selecting the Runner group, File > New > File... Property List
![Screen Shot 2020-06-30 at 3 14 05 PM](https://user-images.githubusercontent.com/682784/86183928-ef1ba700-bae7-11ea-9731-a17259ca7614.png)
1. Call it `Runner.entitlements` (the entitlements extension is important) and add it to the Runner target.
![Screen Shot 2020-06-30 at 3 14 14 PM](https://user-images.githubusercontent.com/682784/86183966-05296780-bae8-11ea-8465-8a8b2cac6d24.png)
1. Add the same Associated Domains you added to the App Clip entitlements file.  DO NOT add the Parent Application Identifiers.
![Screen Shot 2020-06-30 at 3 14 37 PM](https://user-images.githubusercontent.com/682784/86183999-17a3a100-bae8-11ea-8eae-fd298229c2bb.png)
1. Open up your Runner target build settings.  Set `Code Signing Entitlements` to the path of the Runner.entitlements file you just made, like `Runner/Runner.entitlements`.
![Screen Shot 2020-06-30 at 3 15 15 PM](https://user-images.githubusercontent.com/682784/86184092-44f04f00-bae8-11ea-9926-8ce82ea6ee13.png)
1. Open up your App Clip target build settings.  Set `iOS Deployment Target` to iOS 13.0 (or just delete the setting to inherit).
![Screen Shot 2020-06-30 at 3 16 01 PM](https://user-images.githubusercontent.com/682784/86184347-e081bf80-bae8-11ea-861c-f46c1e9ec7c4.png)
1. Still in your App Clip target build settings, in `Framework Search Paths` add `$(inherited)` and `$(PROJECT_DIR)/Flutter` in order so it can find the Flutter framework.  
![Screen Shot 2020-06-30 at 3 16 31 PM](https://user-images.githubusercontent.com/682784/86184451-1aeb5c80-bae9-11ea-8ca3-e2881e700791.png)
1. Still in your App Clip target build settings, set `Objective-C Bridging Header` to `Runner/Runner-Bridging-Header.h`.
1. Now go to your App Clip target Build Phases tab.  Add a New Run Script Phase.
![Screen Shot 2020-06-30 at 3 18 00 PM](https://user-images.githubusercontent.com/682784/86184599-600f8e80-bae9-11ea-8f8c-db2a8391600e.png)
1. Drag it it so it's right below Dependencies, and paste the following (this matches the build phase in your Runner target).
```shell
/bin/sh "$FLUTTER_ROOT/packages/flutter_tools/bin/xcode_backend.sh" build
```
This will compile your Dart code.  This should only happen once when you build your project since the back end will cache the results.
![Screen Shot 2020-06-30 at 3 18 26 PM](https://user-images.githubusercontent.com/682784/86185276-d19c0c80-baea-11ea-981a-13fdcdb3af92.png)
1. Make another New Run Script Phase.  Drag it so it's below Copy Bundle Resources
```shell
/bin/sh "$FLUTTER_ROOT/packages/flutter_tools/bin/xcode_backend.sh" embed_and_thin
```
This will embed your Dart code and the Flutter engine into the App Clip bundle.
![Screen Shot 2020-06-30 at 3 18 46 PM](https://user-images.githubusercontent.com/682784/86185288-d496fd00-baea-11ea-8703-4e4d2fc3a16c.png)
1. Run the App Clip scheme, targeting a iOS 14 device.
![scheme](https://user-images.githubusercontent.com/682784/86185356-0d36d680-baeb-11ea-8661-168553ebcac9.png)

And it worked!  Note the "Powered by" banner at the top.
![Screen Shot 2020-06-30 at 3 21 59 PM](https://user-images.githubusercontent.com/682784/86185404-2e97c280-baeb-11ea-86fa-4574c64de463.png)

### Caveats
1. I make no guarantee of this working for your project, it's just a proof of concept to get you started.  Please don't @ me to fix any linker errors, build setting problems, etc. 🙂
1. Obviously this is all manual, and ideally Flutter tooling would do all this for you.
1. I wasn't able to connect to the observatory, possibly because an entitlement is missing?
1. This will embed your entire app into the App Clip.  You'll need to figure out the right entrypoint (and create a separate AppDelegate for each target, don't share them).  It needs to stay under 10GB per Apple's requirements.
1. This could potentially be solved by embedding two Flutter add-to-app modules into a host app clip target?
1. I didn't have any plugins. Hopefully it would be pretty easy to add the new target to your [`ios/Podfile`](https://guides.cocoapods.org/syntax/podfile.html) but I didn't try, maybe (in master) something like:
```ruby
project 'RunnerAppClip', {
  'Debug' => :debug,
  'Profile' => :release,
  'Release' => :release,
}
...
target 'Runner' do
  flutter_install_all_ios_pods File.dirname(File.realpath(__FILE__))
  target 'RunnerAppClip' do
    inherit! :search_paths
  end
end
```

Here's a zipped up version of that `ios` directory so you can open the Xcode project yourself and drop into a newly created `flutter create test_appclip` demo project.  You'll need to change the signing team.
[ios.zip](https://github.com/flutter/flutter/files/4854670/ios.zip)

**Comment by EliasDeuss** [comment]
@jmagman Wow, thanks for workaround. Also I think you meant 10mb not 10GB.

**Comment by jmagman** [comment]
@EliasDeuss Thanks, updated.

**Comment by xster** [comment]
Seems more or less as simplified as I would expect :D

I think as a compromise between full manual and full tool driven xcode proj manipulation, we can explore a middle ground where the users start a project via `flutter create --sample=ios-app-clip` or some such (or recreate their unmodified existing iOS project). It leaves them with the xcodeproj pre-hooked and 2 vm entrypoints for them to fill. 

> This could potentially be solved by embedding two Flutter add-to-app modules into a host app clip target?

We don't need 2 add-to-app modules (which are platform wrappings of embedding APIs + Dart) right? We just need 2 Dart libraries I think.

**Comment by piotrtobolski** [comment]
Does flutter support multiple app targets already? I think it doesn't but it would be necessary to have it, to correctly support App Clips because App Clip is just another target type.

Implementing App Clips correctly would require supporting all features of multiple targets so there has to be a way to define which assets, plugins etc. belong to which target (because of the App Clip size constraints). I think it would be needed to provide Dart app entrypoint for a target (or some other compile-time way to distinguish if the code is path of which target) so the extraneous Dart code could be automatically stripped thanks to the tree shaking algorithm (if I understand it correctly).

I think the Flutter framework embedded in App Clip would be the same but maybe there are some parts that are heavy but not required for simple apps so we could have slimmed down version that would leave more than 5MB for the rest of the app? In this context every byte is important.

**Comment by xster** [comment]
`flutter create --sample=ios-app-clip`

unfortunately this doesn't work (in its current state). It just applies changes on lib/main.dart https://github.com/flutter/flutter/blob/8bd2e6585b0eda5ccdc10c1e1d35105e11f84424/packages/flutter_tools/lib/src/commands/create.dart#L690.

**Comment by xster** [comment]
I couldn't get observatory attach to work in the app clip either. Even with the correction from https://github.com/flutter/flutter/issues/63893. Unfortunately, I suspect it's intentional. 

Considering
![Screen Shot 2020-09-07 at 9 31 32 PM](https://user-images.githubusercontent.com/156888/92433304-81c46c00-f151-11ea-8c24-821bbf516c36.png)
from https://developer.apple.com/documentation/app_clips/choosing_the_right_functionality_for_your_app_clip, mDNS features might be disabled in app clips. 

As a fallback, what will be needed as a non-ideal workflow is to capture the observatory URI yourself in the console output (such as in Xcode):

![Screen Shot 2020-09-07 at 9 33 37 PM](https://user-images.githubusercontent.com/156888/92433408-d49e2380-f151-11ea-8d44-339aecdc98c3.png)

then re-specify it in the `flutter attach` command, such as `flutter attach --debug-uri http://127.0.0.1:50003/GCk4QIe203U=/`

![Screen Shot 2020-09-07 at 9 34 37 PM](https://user-images.githubusercontent.com/156888/92433439-eed80180-f151-11ea-864c-c1851854c269.png)

In other words, mDNS is just an auto-discovery mechanism. You can still debug by pointing to the running process's port directly.

**Comment by xster** [comment]
The cocoapods part will take some digging in. I couldn't get the "[CP] Embed Pods Frameworks" build phase step to show up automatically from pod install on the app clip target. Adding it manually works but probably need to tweak the Podfile usage a bit.

Additionally, we'll need to make sure Podfile sets bitcode to false on the app clip too.

**Comment by xster** [comment]
Doh, I had to add it manually because it's not implemented in CocoaPods yet. It's in https://github.com/CocoaPods/CocoaPods/commit/6eb5c13dc71fe9fe8452a77813a2e97b1f7245d0 which is in head but only in beta tags 1.10.0.beta.1, 2 so far. Latest stable at the time of writing is 1.9.3.

**Comment by xster** [comment]
Still couldn't figure out the cocoapods part yet. 

Even on HEAD, it doesn't want to add the embed build phase because in https://github.com/CocoaPods/CocoaPods/blob/0d1a93082d8519b4d768bb6c6fedeee472994271/lib/cocoapods/installer/user_project_integrator/target_integrator.rb#L547, the `framework_paths_by_config` is empty for the App Clip target https://github.com/CocoaPods/CocoaPods/blob/0d1a93082d8519b4d768bb6c6fedeee472994271/lib/cocoapods/target/aggregate_target.rb#L245. 

I can't add 

```ruby
project 'RunnerAppClip', {
  'Debug' => :debug,
  'Profile' => :release,
  'Release' => :release,
}
```

either since it's not a separate xcodeproj. I don't really know the Podfile syntax but I assume it's not necessary either since all `target` should inherit from the root level `project` declaration anyway. 

Will have to do more source reading to figure out how cocoapods works.

**Comment by xster** [comment]
I'm not sure I understand the innards of CocoaPods enough but I couldn't get the general structure of 

```ruby
target 'Runner' do
  use_frameworks!
  use_modular_headers!

  flutter_install_all_ios_pods File.dirname(File.realpath(__FILE__))
  target 'AppClip' do
    inherit! :search_paths
  end
end
```

to add an embed frameworks build phase. 

One instead has to do 

```ruby
use_frameworks!
use_modular_headers!
flutter_install_all_ios_pods File.dirname(File.realpath(__FILE__))

target 'Runner'

target 'AppClip'
``` 

which in theory according to https://guides.cocoapods.org/using/the-podfile.html should behave the same but in practice doesn't. The latter produces the list `[#<Pod::PodTarget name=Flutter-iOS14.0>, #<Pod::PodTarget name=device_info-iOS14.0>]` for `pod_targets_for_build_configuration` internally.

For folks with other targets they don't want to pollute, this is equivalent to

```ruby
abstract_target 'TargetsThatUseFlutter' do #this name doesn't matter
  use_frameworks!
  use_modular_headers!
  flutter_install_all_ios_pods File.dirname(File.realpath(__FILE__))

  target 'Runner'

  target 'AppClip'
end
```

**Comment by xster** [comment]
> Last remaining issue. Despite the plugin code being run and being added to the plugin registry, it doesn't handle the incoming method call for some reason. Will have to debug with a local engine.

Never mind, plugins work fine. An orthogonal issue is that we can't use UIWindowSceneDelegate or have the <key>UIApplicationSceneManifest</key> in the Info.plist since our hook in https://github.com/xster/engine/blob/1f3013163d48f46cc967b31aac193691ef28f7d9/shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm#L150 wouldn't work. Just using a bland main storyboard works.

**Comment by xster** [comment]
For the sake of having clear deliverables, I'm scoping this bug to be manual additions of Flutter-powered App Clip targets to an iOS app. For automatic integrations from tools, I'm splitting it off to https://github.com/flutter/flutter/issues/65451.

**Comment by miguelpruivo** [comment]
I'm really struggling to get a build properly done with Flutter + AppClip (native with Swift UI). Clip on its own runs just fine, but Runner target seems to fail no matter what.

Anyone has experienced this?

**Comment by xster** [comment]
@miguelpruivo could you open a new issue for it? Sounds like you're asking for something slightly different than the premise of this bug. i.e. you want to have your App Clip be a standard SwiftUI App Clip rather than a Flutter powered App Clip while your main app is Flutter. Is that right? Could you describe your reproduction steps / project and the exact error?

**Comment by xster** [comment]
FWIW, this is what it's supposed to look like after you install the App Clip on your test phone and then scan a generated QR code following https://developer.apple.com/documentation/app_clips/testing_your_app_clip_s_launch_experience

![Screen Shot 2020-09-26 at 12 02 00 AM](https://user-images.githubusercontent.com/156888/94334493-829c3100-ff8c-11ea-96a1-c006696a090b.png)
![Screen Shot 2020-09-26 at 12 02 05 AM](https://user-images.githubusercontent.com/156888/94334517-8465f480-ff8c-11ea-9d9e-f2796024037e.png)

**Comment by majingAragon** [comment]
I have added app clip into our flutter project and run well, but how to split part of plugins and assets out to app clip target? because app clip size has constraints.

**Comment by miguelpruivo** [comment]
@majingAragon I need to know this as well. Currently clip is inheriting all the Runner dependencies, which makes it bloated and bigger than 10MB.

**Comment by xster** [comment]
@majingAragon can you open another issue describing your issue? Splitting plugins and assets are likely not very useful unless you also split the Dart code referencing those different subsets of plugins and assets (which is currently not supported). 

We're doing some work in that area in https://github.com/flutter/flutter/issues/57617 and you can track our progress there.

**Comment by jimad7** [comment]
When are you going to fully support app clips in a Flutter project? The official docs says that it's possible. I have succeded to create an app clip. But the size of a simple app clip, is way over the 10MB limit from just showing a simple screen. I've also tried to embed a simple swift view, without any shared resources from the main Flutter app. But the Flutter Framework is by default embedded in the app clip target. Therefore i wouldn't say that Flutter supports app clips yet. 
I've really tried everything to keep app clip size down to minimum but it's seems impossible. Am i doing something wrong or are you working to solve this issue?

**Comment by jmagman** [comment]
> I've also tried to embed a simple swift view, without any shared resources from the main Flutter app. But the Flutter Framework is by default embedded in the app clip target.

If you don't have any Flutter elements in your app clip then you shouldn't be embedding any Flutter frameworks in plugins.  There's no "default" here--you have control of what gets embedded in your app clip target.

> But the size of a simple app clip, is way over the 10MB limit from just showing a simple screen.

Out of curiosity, how are you measuring it?

**Comment by jimad7** [comment]
> > I've also tried to embed a simple swift view, without any shared resources from the main Flutter app. But the Flutter Framework is by default embedded in the app clip target.
> 
> If you don't have any Flutter elements in your app clip then you shouldn't be embedding any Flutter frameworks or plugins. There's no "default" here--you have control of what gets embedded in your app clip target.
> 
> > But the size of a simple app clip, is way over the 10MB limit from just showing a simple screen.
> 
> Out of curiosity, how are you measuring it?

Ok, maybe i'm missing something then. I'll try to create the app clip again from scratch and see if i'm able to exclude the Flutter framework in some way. I'm not embedding it all, when archiving the app clip, the framework gets embedded.

I measure my app clip as seen in apple documentation/guidelines. Archive and export as ad hoc or development, app thining and after that checking the app thinning size report text file. When i'm 

If anyone has done this successful (and keeping app clip size below 10MB) please let me know.

**Comment by jerryzhoujw** [comment]
With bitcode = YES,  I can't distribute app with bitcode to apple store.

**Comment by jimad7** [comment]
It's now official that it's not possible to ship an app clip in a Flutter app. The documentation is also updated with this information.

https://github.com/flutter/flutter/issues/71098

And this is a minimal app. My app is way bigger than this (and so is the app clip). Hope they will fix this asap.

**Comment by jmagman** [comment]
Also note you can ship an app clip in a Flutter app, just not a FLUTTER app clip.

**Comment by jimad7** [comment]
> Also note you can ship an app clip in a Flutter app, just not a FLUTTER app clip.

That's interesting. I've tried a native app clip embedded in a Flutter app. No luck, the app clip gets too big. But maybe i'm doing something wrong with the configuration for the app clip target. I would appreciate a small tutorial for that, because i've struggle hours.

**Comment by jmagman** [comment]
You should just be able to open `ios/Runner.xcworkspace`, add the app clip target, and make sure not to reference Flutter at all.  Flutter's pretty much just a few embedded frameworks in a normal iOS app, with a script to generate your compiled dart code, it's not that magic.  So any vanilla non-Flutter App Clip tutorial online should work.  You can dig into the generated App Clip bundle to make sure there's no Flutter pieces sneaking in.
